README.md
## What it does

This package lets you to easily add an action bar over keyboard.

## How to use it

All you need is to wrap your app to `KeyboardActionBarWrapper` and provide `defaultActionBar` to it.

```dart
@override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: KeyboardActionBarWrapper(
        defaultActionBar: _buildDefaultBar,
        child: const MyHomePage(),
      ),
    );
  }

  Widget _buildDefaultBar(FocusNode focusNode) {
    return Container(
      height: 50,
      color: Colors.blue,
      padding: const EdgeInsets.symmetric(horizontal: 20),
      alignment: Alignment.centerRight,
      child: TextButton(
        onPressed: () => focusNode.unfocus(),
        child: const Text(
          'Close',
          style: TextStyle(color: Colors.black),
        ),
      ),
    );
  }
```

After that just use `ActionFocusNode` instead of default `FocusNode`.

```dart
TextField(
  focusNode: ActionFocusNode(),
)
```

## Custom action bar

To use a custom action bar on a specific text field provide `customBar` to its `ActionFocusNode`.

```dart
ActionFocusNode(
  customBar: _buildCustomBar,
);

Widget _buildCustomBar(FocusNode focusNode) {
  final TextTheme textTheme = Theme.of(context).textTheme;

  return Container(
    height: 80,
    color: Colors.red,
    padding: const EdgeInsets.symmetric(horizontal: 20),
    child: Row(
      children: [
        Text(
          'This is custom action bar',
          style: textTheme.bodyText1!.copyWith(color: Colors.white),
        ),
        const Spacer(),
        TextButton(
          onPressed: () => focusNode.unfocus(),
          child: Text(
            'Unfocus',
            style: textTheme.bodyText2!.copyWith(color: Colors.white),
          ),
        ),
      ],
    ),
  );
}
```

## Dismiss on tap outside

To let focus be dismissed on tap outside the keyboard set `unfocusOnTapOutside` to `true`.

```dart
ActionFocusNode(unfocusOnTapOutside: true)
```


CHANGELOG.md
## 1.0.0

Initial release

## 1.0.1

Added hide keyboard by swipe

## 1.0.2

Added barHeight in ActionFocusNode and defaultBarHeight in KeyboardActionBarWrapper

## 1.0.3

Fixed getting the height of the focus node child widget

lib/src/focus_listener.dart
import 'dart:async';

import 'package:chidori/chidori.dart';
import 'package:flutter/material.dart';

class FocusListener {
  final StreamController<FocusEvent> eventPool = StreamController<FocusEvent>.broadcast();

  static final FocusListener _singleton = FocusListener._internal();

  factory FocusListener() {
    return _singleton;
  }

  FocusListener._internal();

  StreamSubscription<FocusEvent> subscribe(Function(FocusEvent event) fn) {
    return eventPool.stream.listen(fn);
  }

  void request(
    ActionFocusNode focusNode, {
    required bool unfocusOnTapOutside,
    Widget Function(ActionFocusNode)? customBar,
  }) {
    eventPool.add(FocusRequested(
      focusNode,
      unfocusOnTapOutside: unfocusOnTapOutside,
      customBar: customBar,
    ));
  }

  void disappear(ActionFocusNode focusNode) {
    eventPool.add(FocusDisappeared(focusNode));
  }
}

class FocusEvent {
  final ActionFocusNode focusNode;

  FocusEvent(this.focusNode);
}

class FocusRequested extends FocusEvent {
  final bool unfocusOnTapOutside;
  final Widget Function(ActionFocusNode)? customBar;

  FocusRequested(
    ActionFocusNode focusNode, {
    this.unfocusOnTapOutside = false,
    this.customBar,
  }) : super(focusNode);
}

class FocusDisappeared extends FocusEvent {
  FocusDisappeared(ActionFocusNode focusNode) : super(focusNode);
}


lib/src/keyboard_action_bar_wrapper.dart
import 'dart:async';
import 'package:chidori/chidori.dart';
import 'package:flutter/material.dart';
import 'package:flutter_keyboard_visibility/flutter_keyboard_visibility.dart';
import 'focus_listener.dart';
import 'package:measured_size/measured_size.dart';

class KeyboardActionBarWrapper extends StatefulWidget {
  final Widget Function(ActionFocusNode) defaultActionBar;
  final Widget child;
  // final double defaultBarHeight;

  const KeyboardActionBarWrapper({
    Key? key,
    required this.defaultActionBar,
    required this.child,
    // this.defaultBarHeight = 50,
  }) : super(key: key);

  @override
  State<KeyboardActionBarWrapper> createState() => _KeyboardActionBarWrapperState();
}

class _KeyboardActionBarWrapperState extends State<KeyboardActionBarWrapper> {
  late final StreamSubscription _subscription;
  late final StreamSubscription _keyboardSubscription;
  ActionFocusNode? currentFocusNode;
  GlobalKey? barKey;

  OverlayEntry? currentEntry;

  bool get _hasFocus => currentFocusNode?.hasFocus ?? false;
  Size wsize = Size.zero;
  double get barHeight {
    return wsize.height;
    // if (currentFocusNode!.barHeight != null) {
    //   return currentFocusNode!.barHeight!;
    // }
    // return widget.defaultBarHeight;
    return 0;
  }

  @override
  void initState() {
    super.initState();
    final keyboardVisibilityController = KeyboardVisibilityController();
    _keyboardSubscription = keyboardVisibilityController.onChange.listen(_handleKeyboardFocus);
    _subscription = FocusListener().subscribe(_handleFocus);
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      margin: EdgeInsets.only(bottom: _hasFocus ? barHeight : 0),
      child: widget.child,
    );
  }

  void _handleFocus(FocusEvent event) {
    if (event is FocusRequested) {
      currentFocusNode = event.focusNode;
      _createEntry(
        unfocusOnTapOutside: event.unfocusOnTapOutside,
        customBar: event.customBar,
      );
    }

    if (event is FocusDisappeared) {
      currentFocusNode = null;
      currentEntry?.remove();
    }

    setState(() {});
  }

  void _handleKeyboardFocus(bool event) {
    if (!event) {
      _unfocus();
    }
  }

  void _createEntry({
    required bool unfocusOnTapOutside,
    Widget Function(ActionFocusNode)? customBar,
  }) {
    if (currentFocusNode == null) {
      return;
    }

    final OverlayState? overlayState = Overlay.of(context);
    if (overlayState == null) {
      return;
    }

    final Widget preBar = customBar?.call(currentFocusNode!) ??
        widget.defaultActionBar(
          currentFocusNode!,
        );

    final bar = MeasuredSize(
      onChange: (Size size) {
        setState(() {
          wsize = size;
        });
      },
      child: preBar,
    );

    currentEntry = OverlayEntry(builder: (context) {
      final MediaQueryData queryData = MediaQuery.of(context);
      return Stack(
        children: [
          if (unfocusOnTapOutside)
            Positioned.fill(
              child: GestureDetector(
                behavior: HitTestBehavior.opaque,
                onTap: _unfocus,
              ),
            ),
          Positioned(
            left: 0,
            right: 0,
            bottom: queryData.viewInsets.bottom,
            child: bar,
          ),
        ],
      );
    });
    overlayState.insert(currentEntry!);
  }

  void _unfocus() {
    currentFocusNode?.unfocus();
  }

  @override
  void dispose() {
    _subscription.cancel();
    _keyboardSubscription.cancel();
    super.dispose();
  }
}


example/lib/main.dart
import 'package:chidori/chidori.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: KeyboardActionBarWrapper(
        defaultActionBar: _buildDefaultBar,
        child: const MyHomePage(title: 'Flutter Demo Home Page'),
      ),
    );
  }

  Widget _buildDefaultBar(FocusNode focusNode) {
    return Container(
      height: 50,
      color: Colors.blue,
      padding: const EdgeInsets.symmetric(horizontal: 20),
      alignment: Alignment.centerRight,
      child: TextButton(
        onPressed: () => focusNode.unfocus(),
        child: const Text(
          'Close',
          style: TextStyle(color: Colors.black),
        ),
      ),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final ActionFocusNode _defaultFocusNode = ActionFocusNode();
  final ActionFocusNode _tapOutsideFocusNode = ActionFocusNode(unfocusOnTapOutside: true);
  late final ActionFocusNode _customFocusNode;

  @override
  void initState() {
    super.initState();
    _customFocusNode = ActionFocusNode(
      customBar: _buildCustomBar,
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Padding(
        padding: const EdgeInsets.symmetric(horizontal: 20),
        child: Column(
          children: [
            Expanded(
              child: SingleChildScrollView(
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    const SizedBox(height: 40),
                    const Text('Default'),
                    TextField(
                      focusNode: _defaultFocusNode,
                    ),
                    const SizedBox(height: 20),
                    const Text('Custom'),
                    TextField(
                      focusNode: _customFocusNode,
                    ),
                    const SizedBox(height: 20),
                    const Text('Dismiss on tap outside'),
                    TextField(
                      focusNode: _tapOutsideFocusNode,
                    ),
                  ],
                ),
              ),
            ),
            Container(
              height: 120,
              width: 200,
              color: Colors.lime,
            ),
          ],
        ),
      ),
    );
  }

  Widget _buildCustomBar(FocusNode focusNode) {
    final TextTheme textTheme = Theme.of(context).textTheme;

    return Container(
      height: 80,
      color: Colors.red,
      padding: const EdgeInsets.symmetric(horizontal: 20),
      child: Row(
        children: [
          Text(
            'This is custom action bar',
            style: textTheme.bodyText1!.copyWith(color: Colors.white),
          ),
          const Spacer(),
          TextButton(
            onPressed: () => focusNode.unfocus(),
            child: Text(
              'Unfocus',
              style: textTheme.bodyText2!.copyWith(color: Colors.white),
            ),
          ),
        ],
      ),
    );
  }
}
